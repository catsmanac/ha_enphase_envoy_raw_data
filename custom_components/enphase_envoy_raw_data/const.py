"""Enphase Envoy Raw Data Support.

This custom integration registers an enphase_envoy_raw_data integration
that only provides a read_data(endpoint) and send_data(endpoint,data)
action/service. No entities are provided.

!!! SENDING DATA TO AN ENVOY ENDPOINT HAS RISK FOR PROPER OPERATION OF THE ENVOY.
DOING SO IS AT YOUR OWN RISK AND SHOULD ONLY BE DONE FULLY UNDERSTANDING ANY EFFECT OF IT !!!

This integration does not replace the core integration. It can be used next to it
"""

from pyenphase import EnvoyAuthenticationError, EnvoyAuthenticationRequired

DOMAIN = "enphase_envoy_raw_data"

CONF_UPDATER = "updater"

NAME = "Enphase Envoy Raw Data"

ENVOY_NAME = "Envoy-raw-data"

UNIQUE_ID = f"{DOMAIN}_for_"

INVALID_AUTH_ERRORS = (EnvoyAuthenticationError, EnvoyAuthenticationRequired)
