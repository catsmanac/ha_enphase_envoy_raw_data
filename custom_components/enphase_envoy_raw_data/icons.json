{
  "services": {
    "read_data": {
      "service": "mdi:download-box-outline"
    },
    "send_data": {
      "service": "mdi:upload-box-outline"
    }
  }
}
