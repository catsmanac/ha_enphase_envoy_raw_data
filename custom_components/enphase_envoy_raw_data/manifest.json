{
  "domain": "enphase_envoy_raw_data",
  "name": "Enphase Envoy Raw Data",
  "codeowners": ["@catsmanac"],
  "config_flow": true,
  "documentation": "https://github.com/catsmanac/ha_enphase_envoy_raw_data#enphase-envoy-raw-data",
  "integration_type": "service",
  "iot_class": "local_polling",
  "loggers": ["pyenphase"],
  "version": "0.0.1",
  "requirements": [
    "pyenphase"
  ]
}
