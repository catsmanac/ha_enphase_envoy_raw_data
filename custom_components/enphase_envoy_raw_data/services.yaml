read_data:
  fields:
    config_entry_id:
      required: true
      selector:
        config_entry:
          integration: enphase_envoy_raw_data
    endpoint:
      required: true
      example: "/"
      selector:
        text:
    from_cache:
      required: false
      example: "false"
      selector:
        boolean:
send_data:
  fields:
    config_entry_id:
      required: true
      selector:
        config_entry:
          integration: enphase_envoy_raw_data
    endpoint:
      required: true
      example: "/"
      selector:
        text:
    data:
      required: true
      selector:
        text:
    risk_acknowledged:
      required: true
      default: false
      selector:
        boolean:
    method:
      required: true
      default: "POST"
      selector:
        select:
          options:
            - "POST"
            - "PUT"
    test_mode:
      required: true
      default: true
      selector:
        boolean:
