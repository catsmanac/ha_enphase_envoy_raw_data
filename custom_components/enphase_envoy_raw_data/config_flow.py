"""Configuration for Enphase Envoy Raw Data Support.

This custom integration registers an enphase_envoy_raw_data integration
that only provides a read_data(endpoint) and send_data(endpoint,data)
action/service. No entities are provided.

!!! SENDING DATA TO AN ENVOY ENDPOINT HAS RISK FOR PROPER OPERATION OF THE ENVOY.
DOING SO IS AT YOUR OWN RISK AND SHOULD ONLY BE DONE FULLY UNDERSTANDING ANY EFFECT OF IT !!!

This integration does not replace the core integration. It can be used next to it
"""

from collections.abc import Mapping
import logging
from typing import Any

from awesomeversion import AwesomeVersion
from pyenphase import AUTH_TOKEN_MIN_VERSION, Envoy, EnvoyError
import voluptuous as vol

from homeassistant.config_entries import (
    SOURCE_REAUTH,
    ConfigEntry,
    ConfigFlow,
    ConfigFlowResult,
)
from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PASSWORD, CONF_USERNAME
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.httpx_client import get_async_client
from homeassistant.helpers.typing import VolDictType

from .const import DOMAIN, ENVOY_NAME, INVALID_AUTH_ERRORS, UNIQUE_ID

type OurConfigEntry = ConfigEntry

CONF_SERIAL = "serial"

INSTALLER_AUTH_USERNAME = "installer"

_LOGGER = logging.getLogger(__name__)


async def validate_input(
    hass: HomeAssistant,
    host: str,
    username: str,
    password: str,
    errors: dict[str, str],
    description_placeholders: dict[str, str],
) -> Envoy:
    """Validate the user input allows us to connect."""
    envoy = Envoy(host, get_async_client(hass, verify_ssl=False))
    try:
        await envoy.setup()
        await envoy.authenticate(username=username, password=password)
    except INVALID_AUTH_ERRORS as e:
        errors["base"] = "invalid_auth"
        description_placeholders["reason"] = str(e)
    except EnvoyError as e:
        errors["base"] = "cannot_connect"
        description_placeholders["reason"] = str(e)
    except Exception:
        _LOGGER.exception("Unexpected exception")
        errors["base"] = "unknown"

    return envoy


class EnphaseExtConfigFlow(ConfigFlow, domain=DOMAIN):
    """Enphase Envoy Raw Datas config flow."""

    VERSION = 1

    def __init__(self) -> None:
        """Initialize an envoy flow."""
        self.ip_address: str | None = None
        self.username = None
        self.protovers: str | None = None

    @callback
    def _async_generate_schema(self) -> vol.Schema:
        """Generate schema."""
        schema: VolDictType = {}

        if self.ip_address:
            schema[vol.Required(CONF_HOST, default=self.ip_address)] = vol.In(
                [self.ip_address]
            )
        elif self.source != SOURCE_REAUTH:
            schema[vol.Required(CONF_HOST)] = str

        default_username = ""
        if (
            not self.username
            and self.protovers
            and AwesomeVersion(self.protovers) < AUTH_TOKEN_MIN_VERSION
        ):
            default_username = INSTALLER_AUTH_USERNAME

        schema[
            vol.Optional(CONF_USERNAME, default=self.username or default_username)
        ] = str
        schema[vol.Optional(CONF_PASSWORD, default="")] = str

        return vol.Schema(schema)

    def _async_envoy_name(self) -> str:
        """Return the name of the envoy."""
        return (
            f"{ENVOY_NAME} {self.unique_id.replace(UNIQUE_ID, '')}"
            if self.unique_id
            else ENVOY_NAME
        )

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the initial step."""
        errors: dict[str, str] = {}
        description_placeholders: dict[str, str] = {}
        host = (user_input or {}).get(CONF_HOST) or self.ip_address or ""

        if user_input is not None:
            envoy = await validate_input(
                self.hass,
                host,
                user_input[CONF_USERNAME],
                user_input[CONF_PASSWORD],
                errors,
                description_placeholders,
            )
            if not errors:
                name = self._async_envoy_name()

                if not self.unique_id:
                    await self.async_set_unique_id(f"{UNIQUE_ID}{envoy.serial_number}")
                    name = self._async_envoy_name()

                if self.unique_id:
                    # If envoy exists in configuration update fields and exit
                    self._abort_if_unique_id_configured(
                        {
                            CONF_HOST: host,
                            CONF_USERNAME: user_input[CONF_USERNAME],
                            CONF_PASSWORD: user_input[CONF_PASSWORD],
                        },
                        error="reauth_successful",
                    )

                # CONF_NAME is still set for legacy backwards compatibility
                return self.async_create_entry(
                    title=name, data={CONF_HOST: host, CONF_NAME: name} | user_input
                )

        if self.unique_id:
            self.context["title_placeholders"] = {
                CONF_SERIAL: self.unique_id.replace(UNIQUE_ID, ""),
                CONF_HOST: host,
            }

        suggested_values: Mapping[str, Any] | None = user_input
        return self.async_show_form(
            step_id="user",
            data_schema=self.add_suggested_values_to_schema(
                self._async_generate_schema(), suggested_values
            ),
            description_placeholders=description_placeholders,
            errors=errors,
        )

    async def async_step_reconfigure(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Add reconfigure step to allow to manually reconfigure a config entry."""
        reconfigure_entry = self._get_reconfigure_entry()
        errors: dict[str, str] = {}
        description_placeholders: dict[str, str] = {}

        if user_input is not None:
            host: str = user_input[CONF_HOST]
            username: str = user_input[CONF_USERNAME]
            password: str = user_input[CONF_PASSWORD]
            envoy = await validate_input(
                self.hass,
                host,
                username,
                password,
                errors,
                description_placeholders,
            )
            if not errors:
                await self.async_set_unique_id(f"{UNIQUE_ID}{envoy.serial_number}")
                self._abort_if_unique_id_mismatch()
                return self.async_update_reload_and_abort(
                    reconfigure_entry,
                    data_updates={
                        CONF_HOST: host,
                        CONF_USERNAME: username,
                        CONF_PASSWORD: password,
                    },
                )

        serial = reconfigure_entry.unique_id.replace(UNIQUE_ID, "") or "-"
        self.context["title_placeholders"] = {
            CONF_SERIAL: serial,
            CONF_HOST: reconfigure_entry.data[CONF_HOST],
        }
        description_placeholders["serial"] = serial

        suggested_values: Mapping[str, Any] = user_input or reconfigure_entry.data
        return self.async_show_form(
            step_id="reconfigure",
            data_schema=self.add_suggested_values_to_schema(
                self._async_generate_schema(), suggested_values
            ),
            description_placeholders=description_placeholders,
            errors=errors,
        )

    async def async_step_reauth(
        self, entry_data: Mapping[str, Any]
    ) -> ConfigFlowResult:
        """Handle configuration by re-auth."""
        return await self.async_step_reauth_confirm()

    async def async_step_reauth_confirm(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Dialog that informs the user that reauth is required."""
        reauth_entry = self._get_reauth_entry()
        errors: dict[str, str] = {}
        description_placeholders: dict[str, str] = {}

        if user_input is not None:
            await validate_input(
                self.hass,
                reauth_entry.data[CONF_HOST],
                user_input[CONF_USERNAME],
                user_input[CONF_PASSWORD],
                errors,
                description_placeholders,
            )
            if not errors:
                return self.async_update_reload_and_abort(
                    reauth_entry,
                    data_updates=user_input,
                )

        serial = reauth_entry.unique_id.replace(UNIQUE_ID, "") or "-"
        self.context["title_placeholders"] = {
            CONF_SERIAL: serial,
            CONF_HOST: reauth_entry.data[CONF_HOST],
        }
        description_placeholders["serial"] = serial
        suggested_values: Mapping[str, Any] = user_input or reauth_entry.data
        return self.async_show_form(
            step_id="reauth_confirm",
            data_schema=self.add_suggested_values_to_schema(
                self._async_generate_schema(), suggested_values
            ),
            description_placeholders=description_placeholders,
            errors=errors,
        )
