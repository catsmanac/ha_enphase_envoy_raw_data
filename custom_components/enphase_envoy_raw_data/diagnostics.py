"""Data Coordinator for Enphase Envoy Raw Data Support.

This custom integration registers an enphase_envoy_raw_data integration
that only provides a read_data(endpoint) and send_data(endpoint,data)
action/service. No entities are provided.

!!! SENDING DATA TO AN ENVOY ENDPOINT HAS RISK FOR PROPER OPERATION OF THE ENVOY.
DOING SO IS AT YOUR OWN RISK AND SHOULD ONLY BE DONE FULLY UNDERSTANDING ANY EFFECT OF IT !!!

This integration does not replace the core integration. It can be used next to it
"""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_NAME,
    CONF_PASSWORD,
    CONF_TOKEN,
    CONF_UNIQUE_ID,
    CONF_USERNAME,
)
from homeassistant.core import HomeAssistant

TO_REDACT = {
    CONF_NAME,
    CONF_PASSWORD,
    # Config entry title and unique ID may contain sensitive data:
    "title",
    CONF_UNIQUE_ID,
    CONF_USERNAME,
    CONF_TOKEN,
}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""

    diagnostic_data: dict[str, Any] = {
        "config_entry": async_redact_data(entry.as_dict(), TO_REDACT),
    }

    return diagnostic_data
