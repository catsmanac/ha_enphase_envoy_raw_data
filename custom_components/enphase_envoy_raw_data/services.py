"""Services for Enphase Envoy Raw Data Support.

This custom integration registers an enphase_envoy_raw_data integration
that only provides a read_data(endpoint) and send_data(endpoint,data)
action/service. No entities are provided.

!!! SENDING DATA TO AN ENVOY ENDPOINT HAS RISK FOR PROPER OPERATION OF THE ENVOY.
DOING SO IS AT YOUR OWN RISK AND SHOULD ONLY BE DONE FULLY UNDERSTANDING ANY EFFECT OF IT !!!

This integration does not replace the core integration. It can be used next to it
"""

from __future__ import annotations

import logging
from typing import Any, Never

from httpx import HTTPError, Response
import orjson
from pyenphase import EnvoyError
from pyenphase.const import URL_TARIFF
import voluptuous as vol

from homeassistant.config_entries import ConfigEntryState
from homeassistant.core import (
    HomeAssistant,
    ServiceCall,
    ServiceResponse,
    SupportsResponse,
)
from homeassistant.exceptions import HomeAssistantError, ServiceValidationError

from .const import DOMAIN
from .coordinator import EnphaseRawDataUpdateCoordinator

ATTR_CONFIG_ENTRY_ID = "config_entry_id"
ATTR_ENDPOINT = "endpoint"
ATTR_DATA = "data"
ATTR_METHOD = "method"
ATTR_RISK_ACKNOWLEDGED = "risk_acknowledged"
ATTR_VALIDATE_MODE = "test_mode"
ATTR_FROM_CACHE = "from_cache"

REQUESTERRORS = (EnvoyError, HTTPError)

_LOGGER = logging.getLogger(__name__)


def _raise_validation(key: str, param: str = "") -> Never:
    """Raise Servicevalidation error."""
    raise ServiceValidationError(
        translation_domain=DOMAIN,
        translation_key=key,
        translation_placeholders={
            "args": param,
        },
    )


def _raise_ha_error(call: ServiceCall, key: str, host: str, param: str) -> Never:
    """Raise HomeAssistant error."""
    raise HomeAssistantError(
        translation_domain=DOMAIN,
        translation_key=key,
        translation_placeholders={
            "host": host,
            "args": f"{call.service} {param}",
        },
    )


def _find_envoy_coordinator(
    hass: HomeAssistant, call: ServiceCall
) -> EnphaseRawDataUpdateCoordinator:
    """Find envoy config entry from service data and return envoy coordinator."""
    identifier = str(call.data.get(ATTR_CONFIG_ENTRY_ID))
    _LOGGER.debug("Finding coordinator for %s", identifier)
    if not (entry := hass.config_entries.async_get_entry(identifier)):
        _raise_validation("envoy_service_no_config", identifier)
    if entry.state is not ConfigEntryState.LOADED:
        _raise_validation("not_initialized", identifier)

    coordinator: EnphaseRawDataUpdateCoordinator = entry.runtime_data
    if (
        not (coordinator)
        or not (envoy_to_use := coordinator.envoy)
        or not envoy_to_use.data
        or not envoy_to_use.data.raw
    ):
        _raise_validation("not_initialized", identifier)
    return coordinator


async def _envoy_request(
    hass: HomeAssistant,
    call: ServiceCall,
    endpoint: str,
    method: str | None = None,
    data: dict[str, Any] | None = None,
    to_cache: bool = False,
    from_cache: bool = False,
) -> Any:
    """Send request to envoy an return reply."""
    coordinator = _find_envoy_coordinator(hass, call)
    envoy_to_use = coordinator.envoy
    if from_cache and endpoint in envoy_to_use.data.raw:
        _LOGGER.debug(
            "envoy_request, return data from cache %s", envoy_to_use.data.raw[endpoint]
        )
        return envoy_to_use.data.raw[endpoint]
    try:
        _LOGGER.debug("envoy_request, sending request to %s", endpoint)
        reply: Response = await envoy_to_use.request(endpoint, data, method)
    except REQUESTERRORS as err:
        _raise_ha_error(call, "envoy_error", envoy_to_use.host, err.args[0])

    if not (200 <= reply.status_code < 300):
        _raise_ha_error(
            call,
            "envoy_error",
            f"{envoy_to_use.host}{endpoint}",
            f"{reply.status_code} {reply.reason_phrase}",
        )
    _LOGGER.debug(
        "envoy_request, request status %s %s", reply.status_code, reply.reason_phrase
    )

    try:
        result = orjson.loads(reply.content)
    except (orjson.JSONDecodeError, ValueError):
        # it's xml or html
        _LOGGER.debug("envoy_request, No JSON data returned, decode it")
        result = reply.content.decode("utf-8")
    if to_cache:
        envoy_to_use.data.raw[endpoint] = result
    return result


async def setup_hass_services(hass: HomeAssistant) -> ServiceResponse:
    """Configure Home Assistant services for Enphase_Envoy."""

    async def read_data_service(call: ServiceCall) -> ServiceResponse:
        """Send GET request to envoy."""
        endpoint = call.data[ATTR_ENDPOINT]
        _LOGGER.debug("read_data_service, reading endpoint %s", endpoint)
        reply = await _envoy_request(
            hass,
            call,
            endpoint=endpoint,
            to_cache=True,
            from_cache=call.data.get(ATTR_FROM_CACHE, False),
        )
        return {endpoint: reply}

    # declare read request services
    hass.services.async_register(
        DOMAIN,
        "read_data",
        read_data_service,
        schema=vol.Schema(
            {
                vol.Required(ATTR_CONFIG_ENTRY_ID): str,
                vol.Required(ATTR_ENDPOINT): str,
                vol.Optional(ATTR_FROM_CACHE): bool,
            }
        ),
        supports_response=SupportsResponse.OPTIONAL,
    )

    async def send_data_service(call: ServiceCall) -> ServiceResponse:
        """Send put or post request to envoy."""
        if not call.data[ATTR_RISK_ACKNOWLEDGED]:
            _raise_validation("not_acknowledged", call.service)
        endpoint = call.data[ATTR_ENDPOINT]
        data = call.data[ATTR_DATA]
        _LOGGER.debug("send_data_service endpoint: %s data: %s", endpoint, data)
        try:
            # make data dict or list
            data_to_send = data if isinstance(data, (dict)) else orjson.loads(str(data))
        except (orjson.JSONDecodeError, ValueError, TypeError) as err:
            _raise_validation(
                "envoy_service_invalid_parameter",
                f", Error: {err.args[0]}, Data: {data}",
            )

        # if in validate mode return data to send
        if call.data.get(ATTR_VALIDATE_MODE):
            try:
                _LOGGER.debug(
                    "send_data_service, test mode, not sending data, returning formatted data: %s",
                    {endpoint: data_to_send},
                )
                return {endpoint: data_to_send}
            except TypeError as err:
                _raise_validation(
                    "envoy_service_invalid_parameter",
                    f", Error: {err.args[0]}, Data: {data}",
                )
        try:
            reply = await _envoy_request(
                hass,
                call,
                endpoint=endpoint,
                method=call.data.get(ATTR_METHOD),
                data=dict(data_to_send),
            )
        except TypeError as err:
            _raise_validation(
                "envoy_service_invalid_parameter",
                f", Error: {err.args[0]}, Data: {data}",
            )

        return {endpoint: reply}

    # declare SEND request services
    hass.services.async_register(
        DOMAIN,
        "send_data",
        send_data_service,
        schema=vol.Schema(
            {
                vol.Required(ATTR_CONFIG_ENTRY_ID): str,
                vol.Required(ATTR_ENDPOINT): str,
                vol.Required(ATTR_DATA): vol.Any(str, object),
                vol.Required(ATTR_METHOD): vol.In(["PUT", "POST"]),
                vol.Required(ATTR_RISK_ACKNOWLEDGED): bool,
                vol.Required(ATTR_VALIDATE_MODE): bool,
            }
        ),
        supports_response=SupportsResponse.OPTIONAL,
    )

    return None
