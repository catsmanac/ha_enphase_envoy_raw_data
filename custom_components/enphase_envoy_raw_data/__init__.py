"""Initialization for Enphase Envoy Raw Data Support.

This custom integration registers an enphase_envoy_raw_data integration
that only provides a read_data(endpoint) and send_data(endpoint,data)
action/service. No entities are provided.

!!! SENDING DATA TO AN ENVOY ENDPOINT HAS RISK FOR PROPER OPERATION OF THE ENVOY.
DOING SO IS AT YOUR OWN RISK AND SHOULD ONLY BE DONE FULLY UNDERSTANDING ANY EFFECT OF IT !!!

This integration does not replace the core integration. It can be used next to it
"""

from __future__ import annotations

import logging

from pyenphase import Envoy

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.httpx_client import get_async_client
from homeassistant.helpers.typing import ConfigType

from .const import DOMAIN, UNIQUE_ID
from .coordinator import EnphaseRawDataConfigEntry, EnphaseRawDataUpdateCoordinator
from .services import setup_hass_services

_LOGGER = logging.getLogger(__name__)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up Enphase Envoy raw data integration."""

    # setup the enphase_envoy_raw_data services
    await setup_hass_services(hass)
    return True


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Enphase Envoy raw data support from a config entry."""

    host = entry.data[CONF_HOST]
    envoy = Envoy(host, get_async_client(hass, verify_ssl=False))
    coordinator = EnphaseRawDataUpdateCoordinator(hass, envoy, entry)

    # wait for one pyenphase data collection cycle to establish communication
    await coordinator.async_config_entry_first_refresh()
    if not entry.unique_id:
        hass.config_entries.async_update_entry(
            entry, unique_id=f"{UNIQUE_ID}{envoy.serial_number}"
        )

    if entry.unique_id.replace(UNIQUE_ID, "") != envoy.serial_number:
        # If the serial number of the device does not match the unique_id
        # of the config entry, it likely means the DHCP lease has expired
        # and the device has been assigned a new IP address. We need to
        # wait for the next discovery to find the device at its new address
        # and update the config entry so we do not mix up devices.
        raise ConfigEntryNotReady(
            translation_domain=DOMAIN,
            translation_key="unexpected_device",
            translation_placeholders={
                "host": host,
                "expected_serial": str(entry.unique_id.replace(UNIQUE_ID, "")),
                "actual_serial": str(envoy.serial_number),
            },
        )

    entry.runtime_data = coordinator

    # Reload entry when it is updated.
    entry.async_on_unload(entry.add_update_listener(async_reload_entry))

    return True


async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Reload the config entry when it changed."""
    await hass.config_entries.async_reload(entry.entry_id)


async def async_unload_entry(
    hass: HomeAssistant, entry: EnphaseRawDataConfigEntry
) -> bool:
    """Unload a config entry."""
    coordinator = entry.runtime_data
    # cancel scheduled functions
    coordinator.async_cancel_token_refresh()
    coordinator.async_cancel_firmware_refresh()
    return True
