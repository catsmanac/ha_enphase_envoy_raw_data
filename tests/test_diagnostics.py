"""Test Enphase Envoy diagnostics."""

from unittest.mock import AsyncMock

from pyenphase.exceptions import EnvoyError
import pytest
from syrupy.assertion import SnapshotAssertion

from custom_components.enphase_envoy_raw_data.const import DOMAIN

from homeassistant.core import HomeAssistant

from . import setup_integration

from pytest_homeassistant_custom_component.common import MockConfigEntry
from pytest_homeassistant_custom_component.components.diagnostics import get_diagnostics_for_config_entry
from pytest_homeassistant_custom_component.typing import ClientSessionGenerator



# Fields to exclude from snapshot as they change each run
TO_EXCLUDE = {
    "id",
    "device_id",
    "via_device_id",
    "last_updated",
    "last_changed",
    "last_reported",
    "created_at",
    "modified_at",
}


def limit_diagnostic_attrs(prop, path) -> bool:
    """Mark attributes to exclude from diagnostic snapshot."""
    return prop in TO_EXCLUDE


async def test_entry_diagnostics(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    hass_client: ClientSessionGenerator,
    mock_envoy: AsyncMock,
    snapshot: SnapshotAssertion,
) -> None:
    """Test config entry diagnostics."""
    await setup_integration(hass, config_entry)
    assert await get_diagnostics_for_config_entry(
        hass, hass_client, config_entry
    ) == snapshot(exclude=limit_diagnostic_attrs)
