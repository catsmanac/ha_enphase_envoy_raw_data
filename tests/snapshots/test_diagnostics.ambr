# serializer version: 1
# name: test_entry_diagnostics
  dict({
    'config_entry': dict({
      'data': dict({
        'host': '1.1.1.1',
        'name': '**REDACTED**',
        'password': '**REDACTED**',
        'token': '**REDACTED**',
        'username': '**REDACTED**',
      }),
      'disabled_by': None,
      'discovery_keys': dict({
      }),
      'domain': 'enphase_envoy_raw_data',
      'entry_id': '45a36e55aaddb2007c5f6602e0c38e72',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'subentries': list([
      ]),
      'title': '**REDACTED**',
      'unique_id': '**REDACTED**',
      'version': 1,
    }),
  })
# ---
